### ISSUE TITLE
[Doctrine\Common\Annotations\AnnotationException] if using drupal console command inside drupal project directory

### ISSUE THREAD

**Issue by marcusx** [hechoendrupal/drupal-console#2581]
If I try a drupal console command inside a drupal webroot I always get this error. Any ideas whats wrong?

`
[Doctrine\Common\Annotations\AnnotationException]
  [Semantical Error] The class "Drupal\Console\Annotation\DrupalCommand" is not annotated with @Annotation. Are you sure this class can be used as annotation? If so, then you need to add @A
  nnotation to the _class_ doc comment of "Drupal\Console\Annotation\DrupalCommand". If it is indeed no annotation, then you need to add @IgnoreAnnotation("DrupalCommand") to the _class_ do
  c comment of class Drupal\Console\Command\Migrate\DebugCommand.
`

**Comment by aimevp** [comment]
I'm experiencing the same problem

**Comment by gergelypap** [comment]
Same here. I love how every RC update causes a totally different bug. :confused:

**Comment by mickaelperrin** [comment]
Can't get DrupalConsole working, I got a near similar error:

    [Doctrine\Common\Annotations\AnnotationException]
    [Semantical Error] The annotation "@Drupal\Console\Annotations\DrupalCommand" in class Drupal\Console\Command\Locale\TranslationStatusCommand does not exist, or cou
  ld not be auto-loaded.

**Comment by jmolivas** [comment]
is @mickaelperrin did you saw this https://github.com/hechoendrupal/DrupalConsole/issues/2770#issuecomment-254130536 ?

**Comment by jmolivas** [comment]
@mickaelperrin 

Are you running DrupalConsole on development mode (by git clone)? 

If yes did you run composer install ?

**Comment by mickaelperrin** [comment]
My composer.json file had the entry `"drupal/console": "~1"`, and it broked on `composer update`.

I also regenerated explicitely the autoload file by running: `composer dump-autoload -o`.

I can see the class `DrupalCommand` in the `autoload_static.php` file : 

    ./autoload_static.php:        'Drupal\\Console\\Annotations\\DrupalCommand' => __DIR__ . '/..' . '/drupal/console/src/Annotations/DrupalCommand.php',

Am I missing something ?

**Comment by mickaelperrin** [comment]
@jmolivas I would love to get some hints to resolve this. Drupal console is still broken on my setup due to this error, even with latest RC9.

**Comment by jmolivas** [comment]
@mickaelperrin what is the error after updating to rc-9?

**Comment by mickaelperrin** [comment]
@jmolivas always the same.

    [Doctrine\Common\Annotations\AnnotationException]
    [Semantical Error] The annotation "@Drupal\Console\Annotations\DrupalCommand" in class Drupal\Console\Command\Locale\TranslationStatusCommand does not exist, or could not be auto-loaded.

I dig into this yesterday and discovered that the AnnotationRegistry config in Bootstrap/Drupal.php 

    AnnotationRegistry::registerLoader([$this->autoload, "loadClass"]);

is reset by multiple calls to 

    AnnotationRegistry::reset();

in Drupal core when the command discovery is triggered.

**Comment by jmolivas** [comment]
@mickaelperrin how did you installed DrupalConsole and the DrupalConsole Launcher?

Did you follow instructions at https://github.com/hechoendrupal/DrupalConsole/#download-as-new-dependency

Which versions of DrupalConsole (on ten site) and DrupalConsole Launcher (global executable) do you have on your machine?

Which OS and PHP versions do you have on your system?

**Comment by mickaelperrin** [comment]
@jmolivas Thx for trying to help.

I didn't change the way to install DrupalConsole, it is a requirement of my `composer.json` file.

    "drupal/console": "^1.0.0",

I use DrupalConsole in a docker environment on a Mac host. I run drupal console by providing an absolute path to `vendor/bin/drupal` with the `root` and `uri` parameters. I don't use the global DrupalConsole Launcher.

I run PHP 7.0.8, and Drupal 8.2.1.



I don't use the global DrupalConsole Launcher.

**Comment by mickaelperrin** [comment]
Here is the stack trace:

     [Doctrine\Common\Annotations\AnnotationException]
      [Semantical Error] The annotation "@Drupal\Console\Annotations\DrupalCommand" in class Drupal\Console\Command\Locale\TranslationStatusCommand does not exist, or could not be auto-loaded.


    Exception trace:
     () at /src/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/AnnotationException.php:54
     Doctrine\Common\Annotations\AnnotationException::semanticalError() at /src/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/DocParser.php:708
     Doctrine\Common\Annotations\DocParser->Annotation() at /src/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/DocParser.php:641
     Doctrine\Common\Annotations\DocParser->Annotations() at /src/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/DocParser.php:334
     Doctrine\Common\Annotations\DocParser->parse() at /src/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/AnnotationReader.php:195
     Doctrine\Common\Annotations\AnnotationReader->getClassAnnotations() at /src/vendor/doctrine/annotations/lib/Doctrine/Common/Annotations/AnnotationReader.php:203
     Doctrine\Common\Annotations\AnnotationReader->getClassAnnotation() at /src/vendor/drupal/console/src/Annotations/DrupalCommandAnnotationReader.php:23
     Drupal\Console\Annotations\DrupalCommandAnnotationReader->readAnnotation() at /src/vendor/drupal/console/src/Utils/AnnotationValidator.php:46
     Drupal\Console\Utils\AnnotationValidator->isValidCommand() at /src/vendor/drupal/console/src/Application.php:140
     Drupal\Console\Application->registerCommands() at /src/vendor/drupal/console/src/Application.php:40
     Drupal\Console\Application->doRun() at /src/vendor/symfony/console/Application.php:124
     Symfony\Component\Console\Application->run() at /src/vendor/drupal/console/bin/drupal.php:69
     require() at /src/vendor/drupal/console/bin/drupal:4

**Comment by jmolivas** [comment]
Not sure what does this mean?
 >I use DrupalConsole in a docker environment on a Mac host. I run drupal console by providing an absolute path to vendor/bin/drupal with the root and uri parameters. I don't use the global DrupalConsole Launcher.

**Comment by jmolivas** [comment]
DrupalConsole must be installed per site. You need to execute for your drupal site root directory.

```
composer require drupal/console:~1.0 \
--prefer-dist \
--optimize-autoloader \
--sort-packages
```

Yuo did this?

**Comment by mickaelperrin** [comment]
To be clear.

My host is a Mac.
I use Docker (for Mac) to setup my development environment.

I just rerun the composer require command you told me. 

    composer require drupal/console:~1.0 \
    --prefer-dist \
    --optimize-autoloader \
    --sort-packages

In my PHP-FPM (7.x) docker container, I switch to my drupal directory root:

    cd /src/web

Then I run my drupalconsole command:
 
    /src/vendor/bin/drupal --uri=www.mydomain.com list

I need to provide an uri parameter as I don't have any default configuration.

My PR #2851 allows me to run the command in any directory by providing the root parameter:

    /src/vendor/bin/drupal --root=/src/web --uri=www.mydomain.com list

**Comment by mickaelperrin** [comment]
If it can help, I successfully got the `list` command working by adding:

   AnnotationRegistry::registerLoader([\Drupal::service('class_loader'), "loadClass"]);

in the registerCommands function loop.

**Comment by jmolivas** [comment]
@mickaelperrin do this works?

```
cd /src/web
vendor/bin/drupal --uri=www.mydomain.com list
```

**Comment by mickaelperrin** [comment]
No, my `vendor` directory is on the same level of the `web` directory.

**Comment by jmolivas** [comment]
Is your site based on drupal-composer project ?

**Comment by jmolivas** [comment]
then try :
```
cd src
vendor/bin/drupal --uri=www.mydomain.com list
```

**Comment by jmolivas** [comment]
Starting on RC-9 having a configuration file containing `root: web` on the site is no longer required. 

You can keep the file but is required to remove that value from your `drupal-site/console/config.yml`

**Comment by jmolivas** [comment]
@mickaelperrin can you try both

```
cd src
vendor/bin/drupal --uri=www.mydomain.com list
```

```
cd src/web
../vendor/bin/drupal --uri=www.mydomain.com list
```

**Comment by mickaelperrin** [comment]
Yes, my site structure is inspired from `drupal-composer` and acquia `blt` project.

Both commands lead to the same error with the `list` or `init` argument.

**Comment by jmolivas** [comment]
@mickaelperrin  I can see what you did on the PR make it work. But still not sure why do not work out of the box.

Can you try executing from your site root directory (the `src` where your `composer.json` file is)
```
composer show | grep drupal/console
```

**Comment by mickaelperrin** [comment]
Result:

     drupal/console
     drupal/console-core
     drupal/console-en

**Comment by jmolivas** [comment]
which versions?

**Comment by mickaelperrin** [comment]
drupal/console                             1.0.0-rc9           The Drupal CLI. A tool to generate boilerplate code, interact with and debug Drupal.
    drupal/console-core                        1.0.0-rc9           Drupal Console Core
    drupal/console-en                          1.0.0-rc8           Drupal Console English Language

**Comment by jmolivas** [comment]
I will test your PR with an standard installation and to make sure does not introduces any issues.

**Comment by mickaelperrin** [comment]
Do you want me to provide a true PR ?

**Comment by jmolivas** [comment]
A true PR?

I can see you already update your PR https://github.com/hechoendrupal/DrupalConsole/pull/2851

**Comment by mickaelperrin** [comment]
@jmolivas PR #2851 is for adding support for the `root` parameter.

I just added #2946 for this issue.

**Comment by jmolivas** [comment]
@mickaelperrin since you mentioned 
> I use Docker (for Mac) to setup my development environment.

There is a chance I can replicate on my local?  
Having access to the repo, docker files to build the machine or the composer.json file.

**Comment by jmolivas** [comment]
@mickaelperrin Thanks for the PRs.

**Comment by mickaelperrin** [comment]
@jmolivas My setup use multiple git repositories with   use of docker overrides. I also thought about sharing with you a minimal docker setup to reproduce the issue.

But I am currently abroad without wifi connection and use my limited 3G data bandwidth to work. I will try to provide you this as soon as I get an unlimited internet access point.

**Comment by jmolivas** [comment]
@mickaelperrin did you have a chance to have your docker configuration in a place so I can get it and have it tested

**Comment by mickaelperrin** [comment]
@jmolivas I tried to reproduce the error with a minimal Drupal install, but as you stated before I was unable to reproduce the error outside my current and first project (new drupaler here :).

My project is localized in French uses multi site with multi domains access, was installed from existing configuration through the config-installer profile, uses nginx fpm and postgresql as services.

I tried with a really simple composer and docker configuration with the Drupal docker image that uses apache and MySQL. With this setup, Drupal console worked correctly.

For the next ten days my internet connection will be very limited in bandwidth (56kb) this is a pain to run a simple composer update, so trying multiple configuration to setup a reproductive docker testing environment won't be possible.

Currently I have introduced my patch in my composer setup, I am not blocked any more.

It would be interesting if the author of the thread and others commenters that got the issue test my patch with their configuration and report if it worked.
